packages/test-cases/cases/bind/sanity/output/wasm-as/Module/module.ts

```diff
@@ -0,0 +1,168 @@
+    const args: InvokeArgs = wrap_invoke_args(
+      method_size,
+      args_size
+    );
+
+    if (args.method == "moduleMethod") {
+      const result = this.moduleMethodWrapped(args.args, env_size);
+      wrap_invoke_result(result);
+      return true;
+    }
+    else if (args.method == "objectMethod") {
+      const result = this.objectMethodWrapped(args.args, env_size);
+      wrap_invoke_result(result);
+      return true;
+    }
+    else if (args.method == "optionalEnvMethod") {
+      const result = this.optionalEnvMethodWrapped(args.args, env_size);
+      wrap_invoke_result(result);
+      return true;
+    }
+    else if (args.method == "if") {
+      const result = this.ifWrapped(args.args, env_size);
+      wrap_invoke_result(result);
+      return true;
+    }
+    else {
+      wrap_invoke_error(
+        `Could not find invoke function "${args.method}"`
+      );
+      return false;
+    }
+  }
+
+  private moduleMethodWrapped(argsBuf: ArrayBuffer, env_size: u32): ArrayBuffer {
+    const args = deserializemoduleMethodArgs(argsBuf);
+
+    const result = this.moduleMethod(
+      {
+        str: args.str,
+        optStr: args.optStr,
+        en: args.en,
+        optEnum: args.optEnum,
+        enumArray: args.enumArray,
+        optEnumArray: args.optEnumArray,
+        map: args.map,
+        mapOfArr: args.mapOfArr,
+        mapOfMap: args.mapOfMap,
+        mapOfObj: args.mapOfObj,
+        mapOfArrOfObj: args.mapOfArrOfObj
+      }
+    );
+    return serializemoduleMethodResult(result);
+  }
+
+  private objectMethodWrapped(argsBuf: ArrayBuffer, env_size: u32): ArrayBuffer {
+    if (env_size == 0) {
+      throw new Error("Environment is not set, and it is required by method 'objectMethod'")
+    }
+    
+    const envBuf = wrap_load_env(env_size);
+    this.setEnv(Types.Env.fromBuffer(envBuf));
+    const args = deserializeobjectMethodArgs(argsBuf);
+
+    const result = this.objectMethod(
+      {
+        object: args.object,
+        optObject: args.optObject,
+        objectArray: args.objectArray,
+        optObjectArray: args.optObjectArray
+      }
+    );
+    return serializeobjectMethodResult(result);
+  }
+
+  private optionalEnvMethodWrapped(argsBuf: ArrayBuffer, env_size: u32): ArrayBuffer {
+    if (env_size > 0) {
+      const envBuf = wrap_load_env(env_size);
+      this.setEnv(Types.Env.fromBuffer(envBuf));
+    }
+    const args = deserializeoptionalEnvMethodArgs(argsBuf);
+
+    const result = this.optionalEnvMethod(
+      {
+        object: args.object,
+        optObject: args.optObject,
+        objectArray: args.objectArray,
+        optObjectArray: args.optObjectArray
+      }
+    );
+    return serializeoptionalEnvMethodResult(result);
+  }
+
+  private ifWrapped(argsBuf: ArrayBuffer, env_size: u32): ArrayBuffer {
+    const args = deserializeifArgs(argsBuf);
+
+    const result = this._if(
+      {
+        _if: args._if
+      }
+    );
+    return serializeifResult(result);
+  }
```

Comment:
To me, these functions should remain "external" and "static", just as they used to be, because they do not require access to the internals of the module instance. Instead they should call into the module instance from the outside. This will help keep the module class small.

Reply:
Yeah but the problem with that was that since the module is now class `module.ifWrapped` method isn't simply a function that InvokeFunction can represent but also require the Module instance as its argument so the function signature changes there and invoke function not only need function but also need object instance for it to work. 

I tried to pass `this.xyzWrapped` directly but it didn't work, I also tried to use closure but they aren't supported. So, I bundled it as we did in PluginModule.

There may be a better way to do this for sure but this is the one that works for me currently.

Reply:
Nevermind I got the solution!